tests/unit/test_spaces.cpp

```diff
@@ -958,35 +998,42 @@ INSTANTIATE_TEST_SUITE_P(FP16OptFuncs, FP16SpacesOptimizationTest,
  * that has different logic than the float32 and float64 reduce functions.
  * For more info, refer to intel's intrinsics guide.
  */
-#ifdef OPT_AVX512_FP16_VL
 class FP16SpacesOptimizationTestAdvanced : public testing::TestWithParam<size_t> {};
 
 TEST_P(FP16SpacesOptimizationTestAdvanced, FP16InnerProductTestAdv) {
-    auto optimization = cpu_features::GetX86Info().features;
-    if (optimization.avx512_fp16 && optimization.avx512vl) {
-        size_t dim = GetParam();
-        float16 v1[dim], v2[dim];
+    auto optimization = getCpuOptimizationFeatures();
+    size_t dim = GetParam();
+    float16 v1[dim], v2[dim];
 
-        std::mt19937 gen(42);
-        std::uniform_real_distribution<> dis(-0.99, 0.99);
+    std::mt19937 gen(42);
+    std::uniform_real_distribution<> dis(-0.99, 0.99);
 
-        _Float16 baseline = 0;
-        for (size_t i = 0; i < dim; i++) {
-            float val1 = (dis(gen));
-            float val2 = (dis(gen));
-            v1[i] = vecsim_types::FP32_to_FP16((val1));
-            v2[i] = vecsim_types::FP32_to_FP16((val2));
+#if defined(CPU_FEATURES_ARCH_AARCH64) && defined(__GNUC__) && (__GNUC__ < 13)
+    // https://github.com/pytorch/executorch/issues/6844
+    __fp16 baseline = 0;
+#else
+    _Float16 baseline = 0;
+#endif
 
-            baseline += static_cast<_Float16>(val1) * static_cast<_Float16>(val2);
-        }
-        baseline = _Float16(1) - baseline;
+    for (size_t i = 0; i < dim; i++) {
+        float val1 = (dis(gen));
+        float val2 = (dis(gen));
+        v1[i] = vecsim_types::FP32_to_FP16((val1));
+        v2[i] = vecsim_types::FP32_to_FP16((val2));
 
-        auto expected_alignment = [](size_t reg_bit_size, size_t dim) {
-            size_t elements_in_reg = reg_bit_size / sizeof(float16) / 8;
-            return (dim % elements_in_reg == 0) ? elements_in_reg * sizeof(float16) : 0;
-        };
+        baseline += static_cast<decltype(baseline)>(val1) * static_cast<decltype(baseline)>(val2);
+    }
+    baseline = decltype(baseline)(1) - baseline;
```

Comment:
I guess this a question for the intel f16 as well, but why is this here and not in a separate function like the other types?

Reply:
fp16 gets pretty inaccurate pretty quickly (especially with big numbers, which are topical with L2). The basic test template is not good enough for testing the 16-bit floats so we made this one